File: Anurag_19_CSE/Python/day29.py
# Link For Problem: https://leetcode.com/problems/maximum-product-of-word-lengths/

class Solution:

    """
        TC : O(n^2)

        SC : O(n)
    """

    def maxProduct(self, words: list[str]) -> int:
        mask: list[int] = []

        for word in words:
            mask.append(self.helper(word))

        ans: int = 0

        for i in range(len(words)):
            for j in range(len(words)):

                if i != j:
                    if not mask[i] & mask[j]:
                        ans = max(ans, len(words[i])*len(words[j]))

        return ans

    def helper(self, s: str) -> int:
        bit: int = 0

        for c in s:
            idx: int = ord(c)-ord('a')
            bit |= 1 << idx

        return bit

File: Anurag_19_CSE/Java/day29.java
//Link For Problem: https://leetcode.com/problems/maximum-product-of-word-lengths/

public class day29 {

    /*
     * TC : O(n^2)
     * 
     * SC : O(n)
     */

    public int maxProduct(String[] words) {
        int n = words.length;
        int mask[] = new int[n];

        for (int i = 0; i < n; i++)
            mask[i] = helper(words[i]);

        int max = 0;

        for (int i = 0; i < n; i++) {
            for (int j = i; j < n; j++) {

                if ((mask[i] & mask[j]) == 0)
                    max = Math.max(max, words[i].length() * words[j].length());
            }
        }

        return max;
    }

    private int helper(String s) {
        int bit = 0;

        for (char c : s.toCharArray()) {
            int index = c - 'a';
            bit |= 1 << index;
        }

        return bit;
    }
}
